src/dmd/dmodule.d

```diff
@@ -623,26 +634,32 @@ extern (C++) final class Module : Package
             {
                 .error(loc, "unable to read module `%s`", toChars());
                 const pkgfile = FileName.combine(FileName.removeExt(srcfile.toString()), package_d);
-                .errorSupplemental(loc, "Expected '%s' or '%s' in one of the following import paths:",
+                .errorSupplemental(loc,
+                    global.path
+                        ? "Expected '%s' or '%s' in one of the following import paths:"
+                        : "Expected '%s' or '%s' in the import paths.",
                     srcfile.toChars(), pkgfile.ptr);
             }
         }
-        if (!global.gag)
+
+        /* Print path
+         */
+        if (global.path)
         {
-            /* Print path
-             */
-            if (global.path)
+            foreach (idx, path; *global.path)
             {
-                foreach (i, p; *global.path)
-                    fprintf(stderr, "import path[%llu] = %s\n", cast(ulong)i, p);
+                const attr = FileName.exists(path);
+                const(char)* err = attr == 2 ? "" :
+                    (attr == 1 ? " (not a directory)" : " (path not found)");
+                .errorSupplemental(loc, "[%llu]: `%s`%s", cast(ulong)idx, path, err);
```

Comment:
errorSupplemental(loc) would be the wrong interface for getting this information out. Try compiling any of the examples with `-verrors=context` for a real comparison between the old and new diagnostic messages.

Reply:
What do you suggest?

Reply:
`errorSupplemental` should remember the previous location and avoid repeating printing the context location when `-verrors=context`. This happens for other `errorSupplemental` errors. I would consider this an improvement to `errorSupplemental` with `-verrors=context` in a subsequent patch.

Reply:
> What do you suggest?

message() is good for verbatim messages, but dmd wants to be different and sends them to stdout instead of stderr.

If it can't be expressed in one line, I'd probably reference a way to dump search paths of the compiler.

Reply:
> > What do you suggest?
> 
> message() is good for verbatim messages, but dmd wants to be different and sends them to stdout instead of stderr.

Well, message() doesn't write to `DiagnosticHandler`. My initial motivation was that import paths were being written directly with `fprintf` and I wanted them to be written to `DiagnosticHandler` so I can use DMD as a library to create a linter. I can make a PR to fix that.

> If it can't be expressed in one line, I'd probably reference a way to dump search paths of the compiler.

What do you mean? A compiler flag to dump the search paths? I think that is not helpful from the user's perspective. If I'm compiling something the build system will probably specify the include paths via command line, and would probably be easier to get them from the build tool. I think the verbosity flag can be used to limit the number of include paths.

Reply:
> > > What do you suggest?
> > 
> > message() is good for verbatim messages, but dmd wants to be different and sends them to stdout instead of stderr.
> 
> Well, message() doesn't write to `DiagnosticHandler`. My initial motivation was that import paths were being written directly with `fprintf` and I wanted them to be written to `DiagnosticHandler` so I can use DMD as a library to create a linter. I can make a PR to fix that.
> 

Maybe, dmd can do what it sees fit here. All dmd.errors entrypoints go through the gcc diagnostic engine in gdc.

> > If it can't be expressed in one line, I'd probably reference a way to dump search paths of the compiler.
> 
> What do you mean? A compiler flag to dump the search paths? I think that is not helpful from the user's perspective. If I'm compiling something the build system will probably specify the include paths via command line, and would probably be easier to get them from the build tool. I think the verbosity flag can be used to limit the number of include paths.

An example of a bad diagnostic error would be:
```
var.error("conflicts with `%s` at %s", other.toPrettyChars(), other.locToChars());
```
A better example that makes good use of errorSupplemental would be:
```
var.error("conflicts with `%s`", other.toPrettyChars());
errorSupplemental(other.loc, "previous definition here");
```

All uses of error and errorSupplemental should take into account that caret/source line printing is default enabled, so you should avoid calling it too repetitively, include newlines in diagnostic messages, or attempt to format the message in a peculiar way such as vertically aligning them, as it will be rendered horrible and broken up by the two source+underscore lines between each message.

We already have an error to the effect of `%d instantiations, recompile with -v to show more` for instantiation traces deeper than ~6.  There could be a similar one here too - in fact, I'm not sure whether `-v` dumps all search paths, it probably should given everything else it does.

Reply:
> All uses of error and errorSupplemental should take into account that caret/source line printing is default enabled, so you should avoid calling it too repetitively, include newlines in diagnostic messages, or attempt to format the message in a peculiar way such as vertically aligning them, as it will be rendered horrible and broken up by the two source+underscore lines between each message.

Right, I understand the purpose now. Although I think we misuse `errorSupplemental` in some places then. We could have an alternative that does the error alignment but doesn't show the carret, just to add additional information.

> We already have an error to the effect of `%d instantiations, recompile with -v to show more` for instantiation traces deeper than ~6. There could be a similar one here too - in fact, I'm not sure whether `-v` dumps all search paths, it probably should given everything else it does.

We can still show the import paths, but only show like 3 or at maximum 6 or something we define. I think not showing them is not practical at all.